miner/worker.go

```diff
@@ -1057,10 +1129,147 @@ func (w *worker) generateWork(params *generateParams) (*types.Block, error) {
 	defer work.discard()
 
 	w.fillTransactions(nil, work, nil)
-	block, _, err := w.engine.FinalizeAndAssemble(w.chain, work.header, work.state, work.txs, work.unclelist(), work.receipts)
+
+	bestWork := work
+	defer bestWork.discard()
+
+	reward := work.state.GetBalance(consensus.SystemAddress)
+	w.bestProposedBlockLock.RLock()
+	if w.bestProposedBlock != nil && w.bestProposedBlockReward.Cmp(reward) > 0 {
+		log.Info("Prefer proposedBlock", "number", bestWork.header.Number, "balance", reward, "proposedReward", w.bestProposedBlockReward, "type", "generateWork")
+		bestWork = w.bestProposedBlock.copy()
+	}
+	w.bestProposedBlockLock.RUnlock()
+	block, _, err := w.engine.FinalizeAndAssemble(w.chain, bestWork.header, bestWork.state, bestWork.txs, bestWork.unclelist(), bestWork.receipts)
 	return block, err
 }
 
+// fillTransactionsProposedBlock retrieves the pending transactions from the txpool and fills them
+// into the given sealing block. The transaction selection and ordering strategy can
+// be customized with the plugin in the future.
+func (w *worker) fillTransactionsProposedBlock(env *environment, block *ProposedBlockArgs) (error, *big.Int) {
+	gasLimit := env.header.GasLimit
+	blockReward := big.NewInt(0)
+	if env.gasPool == nil {
+		env.gasPool = new(core.GasPool).AddGas(gasLimit)
+		if w.chain.Config().IsEuler(env.header.Number) {
+			env.gasPool.SubGas(params.SystemTxsGas * 3)
+		} else {
+			env.gasPool.SubGas(params.SystemTxsGas)
+		}
+	}
+
+	var coalescedLogs []*types.Log
+	// initialize bloom processors
+	processorCapacity := 100
+	bloomProcessors := core.NewAsyncReceiptBloomGenerator(processorCapacity)
+
+	signal := commitInterruptNone
+	for _, tx := range block.txs {
+		// If we don't have enough gas for any further transactions then we're done
+		if env.gasPool.Gas() < params.TxGas {
+			log.Trace("Not enough gas for further transactions", "have", env.gasPool, "want", params.TxGas)
+			signal = commitInterruptOutOfGas
+			break
```

Comment:
If some transactions are lost in case of out of gas, it will be a disaster when existing sandwich attack.